### ISSUE TITLE
Function arguments are evaluated early

### ISSUE THREAD

**Issue by pulquero** [eclipse/rdf4j#1034]
Currently, function arguments are evaluated early. This can cause unexpected results with custom functions and seems to be in-contra to how everything else is evaluated. The "if" function works as it is treated specially as part of the algebra and not as a general function. Late evaluation could easily be achieved by passing Value wrappers that evaluate lazily, to the function.

		for (int i = 0; i < args.size(); i++) {
			argValues[i] = createLazyValue(args.get(i), bindings);
		}
		return function.evaluate(tripleSource.getValueFactory(), argValues);

**Comment by jeenbroekstra** [comment]
Can you elaborate on where you see early evaluation being a problem?

**Comment by pulquero** [comment]
Custom functions that have behaviour similar to 'if', for example if(expr1,value1,expr2,value2,....,defaultValue) where the first exprN to evaluate to true returns valueN else defaultValue. Early evaluation causes all args to be evaluated which could cause the whole evaluation to fail even though those args may never be relevant. An extreme example would be alwaysReturnTrue(argThatIsIgnored) where argThatIsIgnored is something like 'foo' + 1 (i think that should cause an error). The spec says says that args are evaluated early unless otherwise stated but this is in reference to the definitions of the builtin functions ('if' being one of the stated exceptions).

**Comment by barthanssens** [comment]
Related to #716 ?

**Comment by pulquero** [comment]
Not really, similar but not the same type of lateness. There is no dependency between the two issues.
